Google.Api.Gax.Grpc/Rest/RestChannel.cs

```diff
@@ -64,10 +66,24 @@ public RestChannel(RestServiceCollection serviceCollection, string endpoint, Cha
 
         private async Task<HttpResponseMessage> SendAsync<TRequest>(RestMethod restMethod, string host, CallOptions options, TRequest request, CancellationToken cancellationToken)
         {
-            // TODO: Add headers from options.Headers, but intercept the x-goog-api-client header.
+            var headers = options.Headers
+                .Where(mh => !mh.IsBinary)
+                .Where(mh=> mh.Key != VersionHeaderBuilder.HeaderName)
+                .Select(mh => new KeyValuePair<string, string>(mh.Key, mh.Value))
+                .ToList();
+
+            var restVersion = new VersionHeaderBuilder()
+                .AppendDotNetEnvironment()
+                .AppendAssemblyVersion("gapic", GetType())
+                .AppendAssemblyVersion("gax", typeof(CallSettings))
+                .AppendAssemblyVersion("rest", typeof(HttpClient))
+                .ToString();
+
+            headers.Add(new KeyValuePair<string, string>(VersionHeaderBuilder.HeaderName, restVersion));
+
             // Ideally, add the header in the client builder instead of in the ServiceSettingsBase...
             // TODO: Use options. How do we set the timeout for an individual HTTP request? We probably need to create a linked cancellation token.
-            var httpRequest = restMethod.CreateRequest((IMessage) request, host);
+            var httpRequest = restMethod.CreateRequest((IMessage) request, host, headers);
```

Comment:
I'm not sure that I'd put the header addition into RestMethod, given that nothing about the headers depends on the method. If we keep the header addition in this class instead, we don't need to materialize the list - just add all text headers (other than the version) from the options, then add the version header separately.

Reply:
I've moved it but it feels strange to me that there is a method named `CreateRequest` and it creates a request object … except for the headers.